go/vt/mysqlctl/backup.go

```diff
@@ -413,6 +399,94 @@ func backupFiles(mysqld MysqlDaemon, logger logutil.Logger, bh backupstorage.Bac
 	return nil
 }
 
+// backupFile backs up an individual file.
+func backupFile(ctx context.Context, mysqld MysqlDaemon, logger logutil.Logger, bh backupstorage.BackupHandle, fe *FileEntry, name string, hookExtraEnv map[string]string) (err error) {
+	// Open the source file for reading.
+	var source *os.File
+	source, err = fe.open(mysqld.Cnf(), true)
+	if err != nil {
+		return err
+	}
+	defer source.Close()
+
+	// Open the destination file for writing, and a buffer.
+	wc, err := bh.AddFile(ctx, name)
+	if err != nil {
+		return fmt.Errorf("cannot add file: %v", err)
+	}
+	defer func() {
+		if rerr := wc.Close(); rerr != nil {
+			if err != nil {
+				// We already have an error, just log this one.
+				logger.Errorf("failed to close file %v: %v", name, rerr)
+			} else {
+				err = rerr
+			}
+		}
+	}()
+	dst := bufio.NewWriterSize(wc, 2*1024*1024)
+
+	// Create the hasher and the tee on top.
+	hasher := newHasher()
+	tee := io.MultiWriter(dst, hasher)
+
+	// Create the external write filter, if any.
+	var filterOrTee io.Writer
+	var hwc io.WriteCloser
+	var hclose func() (string, error)
+	if *backupStorageFilter != "" {
+		h := hook.NewHook("backup_filter", []string{"-filter", *backupStorageFilter, "-operation", "write"})
+		h.ExtraEnv = hookExtraEnv
+		hwc, hclose, _, err = h.ExecuteAsWriteFilter(tee)
+		if err != nil {
+			return fmt.Errorf("'backup_filter' hook returned error: %v", err)
+		}
+		filterOrTee = hwc
+	} else {
+		filterOrTee = tee
+	}
+
+	// Create the gzip compression filter
```

Comment:
Now that it's possible to plug in their own filter, users may want the option to disable built-in gzip compression. For example, maybe they want to use the filter hook to use a different compression algorithm, or their filter produces encrypted output that will only get bigger if sent through gzip.